### PR TITLE
Initial contribution for workspace sync feature

### PR DESCRIPTION
**Pull request by vparfonov**
Int this initial contribution will be provided:
 
- `dokerfiles/agent/Dockerfile`: based on `alpine:3.11` and  contains `rsync`, `curl`, `openssh` and [`supercronic`](https://github.com/aptible/supercronic) which will replace `cron` functionality, it because `cron` not work in OpenShift ([issue](https://github.com/gliderlabs/docker-alpine/issues/381)) 
- `dokerfiles/agent/cron` task description that will initial backup operation each 10 minutes
- `dokerfiles/agent/scripts` bash scripts for restore and backup operation

- `watcher/watcher.go`: Go program with responsibility to start restore.sh script on startup, parse output and send progress to the client (che-theia) via websocket

- `dokerfiles/storage/Dockerfile` : based on `alpine:3.11` and  contains `rsync`, `openssh`

Signed-off-by: Vitalii Parfonov <vparfono@redhat.com>

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+       ;;
+    esac
+done
+
+#default
+if [[ -z "$TAG" ]]
+then TAG="latest"
+fi
+if [[ -z "$ORGANIZATION" ]]
+then ORGANIZATION="che-incubator"
```
Comment (ibuziuk) [inline build.sh]: @vparfonov let's put it under https://quay.io/organization/eclipse like kubernetes-image-puller

File: README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+## About Eclipse Che Workspace Data Sync
+
+This tooling provide ability to increase I/O performance for a developer workspaces.
```
Comment (ibuziuk) [inline README.md]: `tooling provides the ability`

File: build.sh (hunk 1 above)
Comment (vparfonov) [reply build.sh]: ok, will do

File: dockerfiles/agent/cron/backup-cron-job [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+# run backup.sh as cron task every 10 minutes
+*/10 * * * * /scripts/backup.sh
```
Comment (ibuziuk) [inline dockerfiles/agent/cron/backup-cron-job:9]: in the future would be nice to somehow configure the backup time ()

File: dockerfiles/agent/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+FROM alpine:3.11
+
+ENV USER=user
+ENV UID=12345
+ENV GID=23456
```
Comment (ibuziuk) [inline dockerfiles/agent/Dockerfile]: saw it in https://stackoverflow.com/questions/49955097/how-do-i-add-a-user-when-im-using-alpine-as-a-base-image/55757473#55757473 hopefully we will not have any surprises on OpenShift ;)

**Review by ibuziuk — APPROVED** [review]
LGTM as the initial contribution, just lets' push images to https://quay.io/organization/eclipse

File: README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+==============
```
Comment (sleshchenko) [inline README.md]: Why it's here? Don't headers help to separate one section from another?

File: README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+## Workspace lifecycle:
+
+#### Workspaces Startup phase:
```
Comment (sleshchenko) [inline README.md:7]: I propose the following format, but I'm not sure.

**Workspaces Startup phase:**
- Pods (workspace and data sync) are started in parallel

**Startup data sync phase:**
- Data flow goes from the persistent volume (rsync server) to the ephemeral volume (rsync client)

**Normal workspace usage phase:**
- Containers in the Workspace Pod have full R/W access to the ephemeral volume with restored project sources
- Container (rsync client) in the Workspace Pod have full R/W access to the persistent volume
- Data flow goes from rsync client to rsync server periodically by cron job

**Workspace Shutdown phase:**
- Containers in the Workspace Pod stopped and data flushed to the persistent volume

File: dockerfiles/agent/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+FROM alpine:3.11
+
+ENV USER=user
```
Comment (sleshchenko) [inline dockerfiles/agent/Dockerfile]:
```
ENV USER=user \
    UID=12345 \
    DID=12345
```
?

File: dockerfiles/agent/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+#cron task not work in openshift in case https://github.com/gliderlabs/docker-alpine/issues/381
+#so will used  supercronic https://github.com/aptible/supercronic
+ENV SUPERCRONIC_URL=https://github.com/aptible/supercronic/releases/download/v0.1.9/supercronic-linux-amd64 \
+    SUPERCRONIC=supercronic-linux-amd64 \
+    SUPERCRONIC_SHA1SUM=5ddf8ea26b56d4a7ff6faecdd8966610d5cb9d85
+
+COPY cron/backup-cron-job /etc/crontabs/backup-cron-job
+COPY scripts scripts
+
+# Add user that will be able to start machine-exec-binary but nothing more
```
Comment (sleshchenko) [inline dockerfiles/agent/Dockerfile]: I assume it's just copy/paste mistake

File: common.sh [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+#
+
+IMAGE_AGENT=workspace-data-sync-agent
+IMAGE_STORAGE=workspace-data-sync-storage
+TAG=$1
+ORGANIZATION=$2
+
+function compile() {
+  echo "Compile file sync progress watcher binary from source code"
+  $(GOOS=linux go build -o ./dockerfiles/agent/scripts/watcher ./watcher/watcher.go)
```
Comment (sleshchenko) [inline common.sh:18]: Why not compile binary in docker with the multistage build?

File: dockerfiles/agent/cron/backup-cron-job (hunk 3 above)
Comment (sleshchenko) [reply dockerfiles/agent/cron/backup-cron-job:9]: won't env vars work here? if no - could we copy into container template that will be initialized with env vars values by entrypoint script?

File: dockerfiles/storage/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    && update-ca-certificates \
+    && rm -rf /var/cache/apk/* \
+    && chmod 0550 /.ssh \
+    && sed -i s/root:!/"root:*"/g /etc/shadow \
+    && chmod +x /usr/local/bin/entrypoint.sh
+
+EXPOSE 2222
+ENTRYPOINT [ "/usr/local/bin/entrypoint.sh" ]
+
+
```
Comment (sleshchenko) [inline dockerfiles/storage/Dockerfile]:
```suggestion
```

File: dockerfiles/storage/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+            ca-certificates \
+    && update-ca-certificates \
+    && rm -rf /var/cache/apk/* \
+    && chmod 0550 /.ssh \
+    && sed -i s/root:!/"root:*"/g /etc/shadow \
+    && chmod +x /usr/local/bin/entrypoint.sh
+
+EXPOSE 2222
+ENTRYPOINT [ "/usr/local/bin/entrypoint.sh" ]
+
```
Comment (sleshchenko) [inline dockerfiles/storage/Dockerfile]:
```suggestion
```

File: pkg/progress/progress_watcher.go [hunk 12]
```diff
@@ -0,0 +1,124 @@
+package progress
```
Comment (sleshchenko) [inline pkg/progress/progress_watcher.go:13]: It's better to put license as header If it's going to become Eclipse project. 
See how to automate it for go here https://github.com/devfile/devworkspace-operator/blob/master/Makefile

File: pkg/progress/progress_watcher.go [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+type Watcher struct {
+	progressPercents int
+	exitCode         int
+	buf              *bufio.Reader
+	subscribers      []func(ProgressState)
+	done             chan bool
+	cmd              *exec.Cmd
+}
+
+func CmdWithProgressWatching(name string, arg ...string) (*Watcher, error) {
```
Comment (sleshchenko) [inline pkg/progress/progress_watcher.go:35]: A few lines of documentation won't hurt anyone :-D

```suggestion
// CmdWithProgressWatching starts cmd and returns watchers that allows to track the progress
func CmdWithProgressWatching(name string, arg ...string) (*Watcher, error) {
```

File: pkg/progress/progress_watcher.go [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+	ticker := time.NewTicker(500 * time.Millisecond)
+
+	for {
+		select {
+		case <-ticker.C:
+			{
+				line, err := p.buf.ReadSlice('\r')
+				if err != nil {
+					fmt.Printf("Failed to read the next line. Cause: %s", err)
+					//TODO continue or break here? Should we send ERROR
```
Comment (sleshchenko) [inline pkg/progress/progress_watcher.go]: Please address that TODO

File: pkg/progress/progress_watcher.go [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+func (p *Watcher) Watch() {
+	// track also if exec is not finished to stop automatically
+	go func() {
+		err := p.cmd.Wait()
+
+		exitCode := p.cmd.ProcessState.ExitCode()
+		p.stop(exitCode)
+
+		if err != nil {
+			//TODO Make sure that it's right way to finish the process
```
Comment (sleshchenko) [inline pkg/progress/progress_watcher.go]: Please address that TODO

File: pkg/progress/progress_watcher.go (hunk 12 above)
Comment (vparfonov) [reply pkg/progress/progress_watcher.go:13]: Thanks :)
Add license header manually. Automate checking will be added in separate PR.

File: README.md (hunk 6 above)
Comment (vparfonov) [reply README.md:7]: Fixed

File: pkg/progress/progress_watcher.go (hunk 13 above)
Comment (vparfonov) [reply pkg/progress/progress_watcher.go:35]: Applied

File: README.md (hunk 5 above)
Comment (vparfonov) [reply README.md]: fixed

File: dockerfiles/agent/Dockerfile (hunk 8 above)
Comment (vparfonov) [reply dockerfiles/agent/Dockerfile]: yes, thanks

File: dockerfiles/storage/Dockerfile (hunk 11 above)
Comment (vparfonov) [reply dockerfiles/storage/Dockerfile]: fixed

File: pkg/progress/progress_watcher.go (hunk 14 above)
Comment (vparfonov) [reply pkg/progress/progress_watcher.go]: Fixed, here I think we can continue restore process

File: pkg/progress/progress_watcher.go (hunk 15 above)
Comment (vparfonov) [reply pkg/progress/progress_watcher.go]: added `log.Fatal(err) //fatal will call os.Exit(1)`

File: dockerfiles/agent/cron/backup-cron-job (hunk 3 above)
Comment (vparfonov) [reply dockerfiles/agent/cron/backup-cron-job:9]: Created issue for this
https://github.com/che-incubator/workspace-data-sync/issues/3

File: common.sh (hunk 9 above)
Comment (vparfonov) [reply common.sh:18]: create issue for this:
https://github.com/che-incubator/workspace-data-sync/issues/4

File: dockerfiles/agent/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+#so will used  supercronic https://github.com/aptible/supercronic
+ENV SUPERCRONIC_URL=https://github.com/aptible/supercronic/releases/download/v0.1.9/supercronic-linux-amd64 \
+    SUPERCRONIC=supercronic-linux-amd64 \
+    SUPERCRONIC_SHA1SUM=5ddf8ea26b56d4a7ff6faecdd8966610d5cb9d85
+
+COPY cron/backup-cron-job /etc/crontabs/backup-cron-job
+COPY scripts scripts
+
+# Add user that will be able to start watcher binary but nothing more
+# the result will be propagated then into scratch image
```
Comment (sleshchenko) [inline dockerfiles/agent/Dockerfile:26]: it's not true because you don't use a scratch image but alpine

File: dockerfiles/storage/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+# Copyright (c) 2019-2020 Red Hat, Inc.
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+
+FROM alpine:3.11
+
+COPY  entrypoint.sh /usr/local/bin
```
Comment (sleshchenko) [inline dockerfiles/storage/Dockerfile:12]:
```suggestion
COPY entrypoint.sh /usr/local/bin
```

File: dockerfiles/storage/Dockerfile [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+
+RUN mkdir  /var/run/sshd /.ssh && \
+    # Change permissions to let any arbitrary user
+    for f in  "/etc/passwd" "/.ssh" "/var/run/sshd" ; do \
+      echo "Changing permissions on ${f}" && chgrp -R 0 ${f} && \
+      chmod -R g+rwX ${f}; \
+    done \
+    && apk update \
```
Comment (sleshchenko) [inline dockerfiles/storage/Dockerfile:20]: should we do `apk clean` after installing needed stuff?

File: dockerfiles/storage/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+    && apk upgrade \
```
Comment (sleshchenko) [inline dockerfiles/storage/Dockerfile:21]: Why do we need it?

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**